runner/components/io.siddhi.distribution.metrics.core/src/main/java/io/siddhi/distribution/metrics/prometheus/reporter/impl/PrometheusReporter.java

```diff
@@ -0,0 +1,158 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.siddhi.distribution.metrics.prometheus.reporter.impl;
+
+import com.codahale.metrics.MetricFilter;
+import com.codahale.metrics.MetricRegistry;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.dropwizard.DropwizardExports;
+import io.prometheus.client.exporter.HTTPServer;
+import io.siddhi.core.exception.ConnectionUnavailableException;
+import org.apache.log4j.Logger;
+import org.wso2.carbon.metrics.core.reporter.impl.AbstractReporter;
+
+import java.io.IOException;
+import java.net.BindException;
+import java.net.InetSocketAddress;
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A reporter which outputs measurements to prometheus.
+ */
+public class PrometheusReporter extends AbstractReporter {
+
+    private final MetricRegistry metricRegistry;
+    private final MetricFilter metricFilter;
+    private PrometheusReporter prometheusReporter;
+    private CollectorRegistry collectorRegistry;
+    private HTTPServer server;
+    private String serverURL;
+
+    private static final Logger log = Logger.getLogger(PrometheusReporter.class);
+
+    private PrometheusReporter(String name, MetricRegistry metricRegistry,
+                               MetricFilter metricFilter, String serverURL) {
+        super(name);
+        this.metricRegistry = metricRegistry;
+        this.metricFilter = metricFilter;
+        this.serverURL = serverURL;
+    }
+
+    @Override
+    public void startReporter() {
+
+        prometheusReporter = PrometheusReporter.forRegistry(metricRegistry, serverURL).filter(metricFilter)
+                .convertRatesTo(TimeUnit.SECONDS).convertDurationsTo(TimeUnit.MILLISECONDS).build();
+
+        try {
+            URL target = new URL(serverURL);
+            collectorRegistry = new CollectorRegistry();
+            collectorRegistry.register(new DropwizardExports(metricRegistry));
+
+            initiateServer(target.getHost(), target.getPort());
+            log.info("Prometheus Server has successfully connected at " + serverURL);
+
+        } catch (IOException | ConnectionUnavailableException e) {
+            log.error("Prometheus Server connection failed at " + serverURL, e);
+        }
+    }
+
+    @Override
+    public void stopReporter() {
+        if (prometheusReporter != null) {
+            disconnect();
+            destroy();
+            prometheusReporter.stop();
+            prometheusReporter = null;
+
+        }
+
+    }
+
+    public static PrometheusReporter.Builder forRegistry(MetricRegistry registry, String serverURL) {
+        return new PrometheusReporter.Builder(registry, serverURL);
+    }
+
+    private void initiateServer(String host, int port) throws ConnectionUnavailableException {
```

Comment:
This can be moved to the top-level field?

Reply:
The above comment is addressed by the commit 88d61ec.

Reply:
Catch MalformedUrlException too, else it will not be handled

Reply:
MalformedUrlException is also caught in the above catch statement.